File: ContiguousBinarySubarray.java
import java.util.HashMap;

/** Approach  Running Sum
    Store running sum in a hash map key-sum, value-index of first occurence of sum
    When same sum is encountered, find the length and update maxLength.

    Edge case: eg[0,1,0,1] 
                option 1: store index -1 as sum 0 in hashmap in yh beginning
                option 2: whenevr sum=0 is encountered length of subarray starts from  beginning of array

    Time complexity - O(N), Space Complexity O(N)
 */

 class Solution {
    public int findMaxLength(int[] nums) {
        int max =0;
        if(nums==null || nums.length == 0) return max;
        int sum =0;
        HashMap<Integer, Integer> map = new HashMap<>();
        //update the map wih initial value(sum=0 at index = -1)
        map.put(0,-1);
        for(int i =0;i<nums.length ;i++){

            //set values of running sum iin variable at each index
            if(nums[i]==0) sum--;
            else sum++;

            // check in map - > if new put sum else check maxSum
            // length = i-map.get(sum)
            if(map.containsKey(map.get(sum))){
                max = Math.max(max, i-map.get(sum));
            }
            else map.put(sum,i);
        }
        return max;
    }
}

File: contiguous-binary-subarray.py
'''
Time complexity: O(n)
Space complexity: O(n)
Did this code successfully run on Leetcode : Yes. (#525)
Any problem you faced while coding this : No.
Approach:
We iterate over the array and maintain a cumulative sum. If we encounter 1, we increase the sum by 1 and if we encounter 0, we decrease the sum by 1.
We maintain a hashmap to store the cumulative sum and the index at which it was encountered. If we encounter the same cumulative sum again, we calculate the length of the subarray and update the maxLength.
We don't update the index if the cumulative sum is already present in the hashmap as we want to find the longest subarray.
EdgeCase: We start with a cumulative sum of 0 at index -1 to handle the case when the subarray starts from the beginning of the array.
'''
from typing import List

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        countMap = {}
        cumSum = 0
        maxLength = 0
        countMap[0] = -1
        for idx, n in enumerate(nums):
            if n == 1:
                cumSum+=1
            else:
                cumSum-=1
            if cumSum in countMap:
                maxLength = max(maxLength, idx - countMap[cumSum])
            else:
                countMap[cumSum] = idx
        return maxLength

#driver code
sol = Solution()
print(sol.findMaxLength([0,1])) #2
print(sol.findMaxLength([0,1,0])) #2
print(sol.findMaxLength([0,1,0,1])) #4